### PR TITLE
Promote logs of business impact from DEBUG

### PR DESCRIPTION
**Pull request by loverdos**
to (usually) INFO. The idea is to avoid the verbosity of DEBUG
and yet be able to understand what is going on, especially
regarding the blockchain progress (or lack thereof). The
usefulness and effectiveness of this patch has been verified
while debugging a real case in production.

Also:

- augment `HeaderGasLimitError` with context to aid debugging
- provide two sub-cases for `InvalidBranch`, thus making the
  precise error clearer
- augment `Block`'s `idTag` with extra info to aid debugging
- use `block.idTag` as the default way to identify a block
  when logging

Ref CGKIELE-397

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/domain/BlockHeader.scala [hunk 1]
```diff
@@ -51,8 +51,10 @@ case class BlockHeader(
 
   lazy val hashAsHexString: String = Hex.toHexString(hash.toArray)
 
-  def idTag: String =
-    s"$number: $hashAsHexString"
+  def idTag: String = {
+    val numberHex = number.toString(16)
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/domain/BlockHeader.scala]: Why this change? Most explorers, including our own, show block in decimal.

File: src/main/scala/io/iohk/ethereum/domain/SignedTransaction.scala [hunk 2]
```diff
@@ -113,11 +114,11 @@ case class SignedTransaction (
...
-    s"""SignedTransaction {
-         |tx: $tx
-         |signature: $signature
-         |sender: ${Hex.toHexString(senderAddress.bytes.toArray)}
-         |}""".stripMargin
+    MoreObjects.toStringHelper(this)
+      .add("tx", tx)
+      .add("signature", signature)
+      .add("senderAddress", Hex.toHexString(senderAddress.bytes.toArray))
+      .toString
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/domain/SignedTransaction.scala]: What will this look like when printed? I was also looking into better formatting here, but was hoping for a more automatic solution, here are some ideas: https://stackoverflow.com/questions/15718506/scala-how-to-print-case-classes-like-pretty-printed-tree

Also, do we add the same for `Block(Body/Header)`, `Transaction`, etc.?

**Comment by shmish111** [comment]
I would suggest that if we are interested in these logs then they should be Riemann events rather than logs

**Comment by loverdos** [comment]
@shmish111 Indeed these are events that can help us understand what the application is doing. That they are in the form of logging is a matter of convention and what we had in place. I would like to find all those important events but this can be approached on-demand. Also this PR includes some model changes regarding error data.

File: src/main/scala/io/iohk/ethereum/ledger/Ledger.scala [hunk 3]
```diff
@@ -470,12 +479,20 @@ class LedgerImpl(
...
       block = block,
       getBlockHeaderByHash = getHeaderFromChainOrQueue,
       getNBlocksBack = getNBlocksBackFromChainOrQueue
     )
+
+    result match {
+      case Left(ValidationBeforeExecError(reason)) ⇒
+        log.error(s"[$context/validateBeforeExecution] Block: ${block.idTag}. Reason: ${reason}")
+      case _ ⇒
+    }
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/ledger/Ledger.scala]: Minor, could be more concise: 
```
result.left.foreach { e => logger.error(...) }
```

File: src/main/scala/io/iohk/ethereum/ledger/Ledger.scala [hunk 4]
```diff
@@ -484,13 +501,37 @@ class LedgerImpl(
...
       stateRootHash = stateRootHash,
       receipts = receipts,
       gasUsed = gasUsed
     )
+
+    result match {
+      case Left(error) ⇒
+        log.error(s"[$context/validateAfterExecution] Block: ${block.idTag}. Reason: ${error.reason}")
+      case _ ⇒
+    }
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/ledger/Ledger.scala]: Minor, conciseness.
